Add correct handling of Userspace Binary version numbers to boot sequence

phil-levis opened the pull request:
### Pull Request Overview

This pull request updates credentials checking so that it properly handles userspace binary version numbers. The AppID TRD says that, at boot, if the kernel encounters two processes with the same Application ID or Short ID, it'll boot the one with the higher version number, of if they have the same version number, the one that the schedule runs first. This patch implements this functionality and cleans up some text in the TRD.

It also adds a new checker for testing, in `kernel::process_checker::basic`, called `AppCheckerRsaSimulated`. This checker acts like a Checking Policy that accepts RSA signatures, assigns Application IDs as the public key of the signature, and short IDs derived from high order bits of the public key (if 0 is the MSB, the short ID is bits 1-31). However, it does not actually check the signature itself, instead blindly accepting an RSA signature. This checker allows you to easily check ID collisions and therefore test version numbers.

### Testing Strategy

This pull request was tested by installing two applications with RSA credentials using the same public key and different version numbers. The kernel correctly only booted the one with the higher version number. If this process is terminated at the process console, the one with the lower version number could successfully be booted from the console.

It was also tested by loading two applications with RSA credentials using the same public key and the same version number. It checks that the kernel runs one of them, whichever the scheduler runs first.

This adds 98 bytes to the kernel:
  - 30 of the bytes are adding an accessor method for application version number to ProcessStandard
  - 52 bytes added to kernel_loop (inlined from loading processes) to handle version numbers
  - 12 bytes because terminate has two extra cases, allowing you to terminate Faulted and Approved apps. You can terminate Faulted processes so that you can restart them; it is not valid to push a stack frame (init task) onto a Faulted process, as it might be storing fault state. So you must first transition it to Terminated. You can terminate CredentialsApproved apps because they are scanned in the main kernel loop to see if they can start. If the kernel wants to check credentials but then not start the process it needs to call terminate() after approving credentials.
  - The remaining bytes are due to alignment noise (+2 here or there).


### TODO or Help Wanted

Feedback on process `State` names.

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

This also adds a lengthy comment to `process.rs` that explains the process state machine.

### Formatting

- [X] Ran `make prepush`.

phil-levis commented:
This adds 98 bytes to the kernel
  - 30 of the bytes are adding an accessor method for application version number to `ProcessStandard`
  - 52 bytes added to `kernel_loop` (inlined from loading processes) to handle version numbers
  - 12 bytes because `terminate` has two extra cases, allowing you to terminate Faulted and Approved apps. You can terminate Faulted processes so that you can restart them; it is not valid to push a stack frame (init task) onto a Faulted process, as it might be storing fault state. So you must first transition it to Terminated. You can terminate CredentialsApproved apps because they are scanned in the main kernel loop to see if they can start. If the kernel wants to check credentials but then not start the process it needs to call terminate() after approving credentials.
  - The remaining bytes are due to alignment noise (+2 here or there).

hudson-ayers commented:
I consider that size increase acceptable

bradjc commented on kernel/src/process.rs:
```diff
@@ -771,8 +776,44 @@ impl From<Error> for ErrorCode {
+/// When a process faults, it enters the `Faulted` state. To be
+/// restarted, it must first transition to the `Terminated` state,
+/// which means that all of its state has been cleaned up.
+
```
```suggestion
```

bradjc commented on kernel/src/process.rs:
```diff
@@ -771,8 +776,44 @@ impl From<Error> for ErrorCode {
 ///
-/// This is made public in case external implementations of `Process` want
-/// to re-use these process states in the external implementation.
+/// This is public so external implementations of `Process` want to
```
```suggestion
/// This is public so external implementations of `Process` can
```

bradjc commented on kernel/src/process_checker.rs:
```diff
@@ -58,40 +60,81 @@ impl<'a> AppCredentialsChecker<'a> for () {
+///   all other processes, or
+///   - Other processes that share an Application Identifier or Short
+///   ID with it have a lower version number and are not running, or
+///   - Other processes that shared
```
?

phil-levis replied:
Looks like it was truncated. Filled it in.

phil-levis commented:
@bradjc -- I think I addressed your comments.

lschuermann commented on doc/reference/trd-appid.md:
```diff
@@ -379,50 +379,76 @@ deterministic behavior in the presence of colliding Application
+process in the `CredentialsApproved` state, it determines whether to
+run the process based on its Application Identifier, Short ID, and the
+Application Binary version number (stored in the Program Header,
+described in Section 5.1). At boot, the kernel starts a process if:
```
```suggestion
described in Section 5.1). At boot, the kernel starts a process if
either of:
```

A nit, just to make it very explicit that this is ORed.

lschuermann commented on doc/reference/trd-appid.md:
```diff
@@ -379,50 +379,76 @@ deterministic behavior in the presence of colliding Application
 Tock process state machine and names of its states is outside the
 scope of this document.
 
+One a Tock system is running, management interfaces may change the set
```
```suggestion
Once a Tock system is running, management interfaces may change the set
```

phil-levis replied:
Done.

hudson-ayers commented on doc/reference/trd-appid.md:
```diff
@@ -379,50 +379,76 @@ deterministic behavior in the presence of colliding Application
+  - B has a version number of 2, an Application ID of Y and a Short ID of 6
+  - C has a version number of 3, an APplication ID of Y and a Short ID of 7
+
+Following the above logic, A cannot run because it shares a Short ID
+with A and has a lower version number. At the same time, B cannot run
```
```suggestion
Following the above logic, A cannot run because it shares a Short ID
with B and has a lower version number. At the same time, B cannot run
```

alexandruradovici commented on doc/reference/trd-appid.md:
```diff
@@ -379,50 +379,76 @@ deterministic behavior in the presence of colliding Application
 Tock process state machine and names of its states is outside the
 scope of this document.
 
+One a Tock system is running, management interfaces may change the set
```
```suggestion
Once a Tock system is running, management interfaces may change the set
```

brghena commented on kernel/src/process_standard.rs:
```diff
@@ -448,7 +453,10 @@ impl<C: Chip> Process for ProcessStandard<'_, C> {
 
     fn terminate(&self, completion_code: Option<u32>) {
-        if !self.is_running() {
+        if !self.is_running()
+            && self.get_state() != State::Faulted
+            && self.get_state() != State::CredentialsApproved
+        {
```
Can you add a comment here explaining why these three states are all part of the early return? I think before it made sense to not terminate a process that isn't running anyways. Now I don't find the logic obvious though.

phil-levis replied:
Great catch. I've added a comment.

hudson-ayers commented on doc/reference/trd-appid.md:
```diff
@@ -379,50 +379,76 @@ deterministic behavior in the presence of colliding Application
+that blocks it does not run. For example, suppose there are three
+processes: A, B and C.
+
+  - A has a version number of 1, an Application ID of X and a Short ID of 6
+  - B has a version number of 2, an Application ID of Y and a Short ID of 6
+  - C has a version number of 3, an APplication ID of Y and a Short ID of 7
```
This suggests that both of the following are true:

1. A given AppID can map to multipl shortIDs
2. A given shortID can map to multip AppIDs

That is a little surprising to me

hudson-ayers commented on kernel/src/process_standard.rs:
```diff
@@ -448,7 +453,10 @@ impl<C: Chip> Process for ProcessStandard<'_, C> {
-        if !self.is_running() {
+        if !self.is_running()
+            && self.get_state() != State::Faulted
+            && self.get_state() != State::CredentialsApproved
```
Why can we not terminate a function that has had its credentials approved, but has not yet run?

phil-levis replied:
Oh, we *can* terminate it. Note the !=. The logic here is confusing, I agree. @brghena pointed this out. I have added a comment to explain.

phil-levis commented:
Alright, I think I've addressed the comments; I explained the logic in `terminate` and removed the incorrect text in the TRD about transitive blocking. Thank you, @hudson-ayers and @brghena for the detailed read-through.

phil-levis commented:
bors r+
